File: frontend/src/types/ShopConfig.ts
export interface ShopConfig {
  shopId: string;
  /** When false, shop landing and game URLs return nothing (Twirla-side offboarding). Omitted defaults to true. */
  enabled?: boolean;
  /** Public URL slug for /shop/:slug and /:mode/:slug/:id */
  slug?: string;
  /** Display name from admin / JSON (optional; falls back to branding.brandName). */
  name?: string;
  adminToken?: string;
  playCooldownHours?: number;
  mode?: ExperienceMode;
  branding: BrandingConfig;
  text: TextConfig;
  cta: CtaConfig;
  wheel?: WheelConfig;
  tapHearts?: TapHeartsConfig;
  scratch?: ScratchConfig;
  countdown?: CountdownConfig;
  /** Flip cards to match pairs; MVP uses emoji/symbol deck */
  memory?: MemoryMatchConfig;
  /** Runner (/runner/...) — weighted end-of-round prize copy */
  runnerGame?: RunnerGameSettings;
  /** Public campaign landing content (shops.json only — no hardcoded shops). */
  campaign?: import('./ShopCampaign').ShopCampaignPageConfig;
}

export enum ExperienceMode {
  Runner = 'Runner',
  Wheel = 'Wheel',
  TapHearts = 'TapHearts',
  Scratch = 'Scratch',
  Countdown = 'Countdown',
  MemoryMatch = 'MemoryMatch',
}

/** Soft drifting shapes behind the experience card (opt-in; keep density low). */
export interface AmbientParticlesConfig {
  /** When true, shows a small number of particles. Omitted or false = off. */
  enabled?: boolean;
  /** Default "orb" — soft blob; "dot" — tiny specks; "ring" — hollow circles. */
  shape?: 'dot' | 'orb' | 'ring';
  /** Default "few" (~6); "some" is still restrained (~10). */
  density?: 'few' | 'some';
  /** Optional CSS color (e.g. hex). Defaults to a wash of brand primary/secondary. */
  color?: string;
  /** Optional second tone for variety. */
  accentColor?: string;
}

export interface ThemeConfig {
  backgroundPattern?: 'gradient' | 'aurora' | 'mesh' | 'dark';
  surfaceStyle?: 'glass' | 'solid' | 'neon';
  /** Optional motion on background orbs (accessibility: use "none" to reduce movement). */
  ambientMotion?: 'none' | 'drift' | 'pulse' | 'shimmer';
  /** Optional very subtle floating particles (brand-colored; respects reduced motion). */
  ambientParticles?: AmbientParticlesConfig;
  fontFamily?: string;
  borderRadius?: number;
  buttonRadius?: number;
}

export interface BrandingConfig {
  primaryColor: string;
  secondaryColor: string;
  logoUrl?: string;
  brandName?: string;
  accentColor?: string;
  theme?: ThemeConfig;
}

export interface TextConfig {
  title: string;
  subtitle?: string;
  ctaText: string;
  resultTitle: string;
  resultSubtitle?: string;
}

export interface CtaConfig {
  url: string;
}

export interface WheelConfig {
  allowRepeatSpins: boolean;
  prizes: PrizeConfig[];
}

export interface PrizeConfig {
  label: string;
  weight: number;
  iconUrl?: string;
  description?: string;
  isWinning?: boolean;
}

/** Weighted random result when Catch the Prize ends (replaces score-based tiers). */
export interface TapHeartsOutcome {
  headline: string;
  description?: string;
  weight: number;
  /** True = no discount / no prize this round. */
  isNoWin?: boolean;
}

export interface TapHeartsConfig {
  heartsToTap: number;
  heartColor: string;
  revealText: string;
  revealSubtitle?: string;
  /** If set, one row is picked at random by weight when the round ends. If omitted, built-in defaults apply. */
  outcomes?: TapHeartsOutcome[];
}

/** Weighted random result when Runner mode ends. */
export interface RunnerGameOutcome {
  headline: string;
  body?: string;
  weight: number;
  isNoWin?: boolean;
}

export interface RunnerGameSettings {
  outcomes?: RunnerGameOutcome[];
}

export interface ScratchConfig {
  overlayColor: string;
  overlayText: string;
  revealText: string;
  revealSubtitle?: string;
}

export interface CountdownConfig {
  endAt: string;
  endMessage: string;
  showCtaBeforeEnd: boolean;
}

export interface MemoryMatchConfig {
  /** Number of pairs (3–8). Total cards = 2 × pairCount */
  pairCount: number;
  /** Shown when all pairs are found */
  revealText: string;
  revealSubtitle?: string;
  /** Optional labels for pairs (length must equal pairCount); default emoji deck if omitted */
  pairLabels?: string[];
  /** Seconds to match all pairs from “Start”; omit for no time limit */
  timeLimitSeconds?: number;
  /** Game ends after this many wrong pairs; omit for unlimited */
  maxMistakes?: number;
}

File: backend/Twirla.Domain/Entities/ShopConfig.cs
using System.Linq;

namespace Twirla.Domain.Entities;

public class ShopConfig
{
    public string ShopId { get; set; } = string.Empty;
    /// <summary>When false, API does not serve this shop (landing/games). Omitted or null defaults to on.</summary>
    public bool? Enabled { get; set; }
    public string? Slug { get; set; }
    public string? Name { get; set; }
    public string? AdminToken { get; set; }
    public int? PlayCooldownHours { get; set; } = 24;
    public BrandingConfig Branding { get; set; } = new();
    public TextConfig Text { get; set; } = new();
    public CtaConfig Cta { get; set; } = new();
    public WheelConfig? Wheel { get; set; }
    public TapHeartsConfig? TapHearts { get; set; }
    public ScratchConfig? Scratch { get; set; }
    public CountdownConfig? Countdown { get; set; }

    public ExperienceMode? GetDefaultMode()
    {
        if (Wheel != null) return ExperienceMode.Wheel;
        if (TapHearts != null) return ExperienceMode.TapHearts;
        if (Scratch != null) return ExperienceMode.Scratch;
        if (Countdown != null) return ExperienceMode.Countdown;
        return null;
    }

    public ShopConfig GetForLanguage(string? language)
    {
        if (string.IsNullOrEmpty(language))
            return this;

        return new ShopConfig
        {
            ShopId = ShopId,
            Enabled = Enabled,
            PlayCooldownHours = PlayCooldownHours,
            Branding = Branding,
            Text = Text.GetForLanguage(language),
            Cta = Cta,
            Wheel = Wheel != null ? new WheelConfig
            {
                AllowRepeatSpins = Wheel.AllowRepeatSpins,
                Prizes = Wheel.Prizes.Select(p => p.GetForLanguage(language)).ToList()
            } : null,
            TapHearts = TapHearts?.GetForLanguage(language),
            Scratch = Scratch?.GetForLanguage(language),
            Countdown = Countdown?.GetForLanguage(language)
        };
    }
}

public enum ExperienceMode
{
    Wheel,
    TapHearts,
    Scratch,
    Countdown
}

public class BrandingConfig
{
    public string PrimaryColor { get; set; } = "#000000";
    public string SecondaryColor { get; set; } = "#FFFFFF";
    public string? LogoUrl { get; set; }
    public string? BrandName { get; set; }
    public string? AccentColor { get; set; }
    public ThemeConfig? Theme { get; set; }
}

public class ThemeConfig
{
    public string? BackgroundPattern { get; set; }
    public string? SurfaceStyle { get; set; }
    /// <summary>Optional motion layer on experience backgrounds (none, drift, pulse, shimmer).</summary>
    public string? AmbientMotion { get; set; }
    public string? FontFamily { get; set; }
    public int? BorderRadius { get; set; }
    public int? ButtonRadius { get; set; }
}

public class TextConfig
{
    public string Title { get; set; } = string.Empty;
    public string? Subtitle { get; set; }
    public string CtaText { get; set; } = string.Empty;
    public string ResultTitle { get; set; } = string.Empty;
    public string? ResultSubtitle { get; set; }
    public Dictionary<string, TextConfig>? Translations { get; set; }

    public TextConfig GetForLanguage(string? language)
    {
        if (string.IsNullOrEmpty(language) || Translations == null || !Translations.ContainsKey(language))
            return this;

        var translated = Translations[language];
        return new TextConfig
        {
            Title = !string.IsNullOrEmpty(translated.Title) ? translated.Title : Title,
            Subtitle = !string.IsNullOrEmpty(translated.Subtitle) ? translated.Subtitle : Subtitle,
            CtaText = !string.IsNullOrEmpty(translated.CtaText) ? translated.CtaText : CtaText,
            ResultTitle = !string.IsNullOrEmpty(translated.ResultTitle) ? translated.ResultTitle : ResultTitle,
            ResultSubtitle = !string.IsNullOrEmpty(translated.ResultSubtitle) ? translated.ResultSubtitle : ResultSubtitle
        };
    }
}

public class CtaConfig
{
    public string Url { get; set; } = string.Empty;
}

public class WheelConfig
{
    public bool AllowRepeatSpins { get; set; }
    public List<PrizeConfig> Prizes { get; set; } = new();
}

public class PrizeConfig
{
    public string Label { get; set; } = string.Empty;
    public int Weight { get; set; }
    public string? IconUrl { get; set; }
    public string? Description { get; set; }
    public bool? IsWinning { get; set; }
    public Dictionary<string, PrizeConfig>? Translations { get; set; }

    public PrizeConfig GetForLanguage(string? language)
    {
        if (string.IsNullOrEmpty(language) || Translations == null || !Translations.ContainsKey(language))
            return this;

        var translated = Translations[language];
        return new PrizeConfig
        {
            Label = !string.IsNullOrEmpty(translated.Label) ? translated.Label : Label,
            Weight = Weight,
            IconUrl = IconUrl,
            Description = !string.IsNullOrEmpty(translated.Description) ? translated.Description : Description,
            IsWinning = IsWinning
        };
    }

    public bool GetIsWinning()
    {
        if (IsWinning.HasValue)
            return IsWinning.Value;

        var losingPatterns = new[] { "try again", "better luck", "no prize", "nothing", "unlucky", "sorry", "no win", "not this time" };
        var labelLower = Label.ToLowerInvariant().Trim();
        return !losingPatterns.Contains(labelLower);
    }
}

public class TapHeartsConfig
{
    public int HeartsToTap { get; set; }
    public string HeartColor { get; set; } = "#FF0000";
    public string RevealText { get; set; } = string.Empty;
    public string? RevealSubtitle { get; set; }
    public Dictionary<string, TapHeartsConfig>? Translations { get; set; }

    public TapHeartsConfig GetForLanguage(string? language)
    {
        if (string.IsNullOrEmpty(language) || Translations == null || !Translations.ContainsKey(language))
            return this;

        var translated = Translations[language];
        return new TapHeartsConfig
        {
            HeartsToTap = HeartsToTap,
            HeartColor = HeartColor,
            RevealText = !string.IsNullOrEmpty(translated.RevealText) ? translated.RevealText : RevealText,
            RevealSubtitle = !string.IsNullOrEmpty(translated.RevealSubtitle) ? translated.RevealSubtitle : RevealSubtitle
        };
    }
}

public class ScratchConfig
{
    public string OverlayColor { get; set; } = "#CCCCCC";
    public string OverlayText { get; set; } = string.Empty;
    public string RevealText { get; set; } = string.Empty;
    public string? RevealSubtitle { get; set; }
    public Dictionary<string, ScratchConfig>? Translations { get; set; }

    public ScratchConfig GetForLanguage(string? language)
    {
        if (string.IsNullOrEmpty(language) || Translations == null || !Translations.ContainsKey(language))
            return this;

        var translated = Translations[language];
        return new ScratchConfig
        {
            OverlayColor = OverlayColor,
            OverlayText = !string.IsNullOrEmpty(translated.OverlayText) ? translated.OverlayText : OverlayText,
            RevealText = !string.IsNullOrEmpty(translated.RevealText) ? translated.RevealText : RevealText,
            RevealSubtitle = !string.IsNullOrEmpty(translated.RevealSubtitle) ? translated.RevealSubtitle : RevealSubtitle
        };
    }
}

public class CountdownConfig
{
    public string EndAt { get; set; } = string.Empty;
    public string EndMessage { get; set; } = string.Empty;
    public bool ShowCtaBeforeEnd { get; set; } = true;
    public Dictionary<string, CountdownConfig>? Translations { get; set; }

    public CountdownConfig GetForLanguage(string? language)
    {
        if (string.IsNullOrEmpty(language) || Translations == null || !Translations.ContainsKey(language))
            return this;

        var translated = Translations[language];
        return new CountdownConfig
        {
            EndAt = EndAt,
            EndMessage = !string.IsNullOrEmpty(translated.EndMessage) ? translated.EndMessage : EndMessage,
            ShowCtaBeforeEnd = ShowCtaBeforeEnd
        };
    }
}
